src/Chainweb/Storage/Table.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- |
+-- Module: Chainweb.Storage.Table
+-- Copyright: Copyright © 2019 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>, Edmund Noble <edmund@kadena.io>
+-- Stability: experimental
+-- Description: Key Value Store
+--
+-- API for Key-Value Stores
+module Chainweb.Storage.Table
+  ( IsCasValue (..)
+  , ReadableTable (..)
+  , ReadableTable1
+  , ReadableCas
+  , Table (..)
+  , Table1
+  , Casify(..)
+  , Cas
+  , casInsert
+  , casInsertBatch
+  , casDelete
+  , casDeleteBatch
+  , IterableTable (..)
+  , IterableTable1
+  , IterableCas
+  , Entry (..)
+  , Iterator (..)
+  , Iterator1
+  , CasIterator
+  , tableLookupM
+  , casLookupM
+  , TableException (..)
+  )
+where
+
+import Control.Exception (Exception, SomeException)
+import Control.Monad.Catch (throwM)
+
+import Data.Coerce
+import Data.Foldable
+import Data.Maybe
+import Data.Text (Text)
+
+import GHC.Generics
+import GHC.Stack
+
+-- | The class of content-addressable values.
+--
+-- The casKey function must be morally injective:
+--
+-- prop> casKey a /= casKey b || a == b
+--
+-- Usually, 'casKey' is a cryptographic, i.e. collision resistant, hash
+-- function.
+class Eq (CasKeyType v) => IsCasValue v where
+    type CasKeyType v
+    casKey :: v -> CasKeyType v
+
+-- | Read-Only View of a Content Addressed Key-Value Store
+--
+-- Since the key uniquely determines the content of the store a value for a key
+-- is either available or not available. There is no dispute about the value
+-- itself.
+class ReadableTable t k v | t -> k v where
+    tableLookup :: t -> k -> IO (Maybe v)
+    tableLookupBatch :: t -> [k] -> IO [Maybe v]
```

Comment:
Instead of `t -> [k] -> IO [Maybe v]`, we could consider `Traversable f => t -> f k -> IO (f (Maybe v))`. 

Not only does this make it more ergonomic to use `tableLookupBatch` with other containers, but it is also a simple way to express at the type level that the number of results returned by this function is guaranteed to be equal to the number of results passed in, since `Traversable` can't change a container's shape.

An interesting variation is to go one step further and have the type:
```haskell
tableLookupBatch :: t -> (forall f. Applicative f => (k -> f (Maybe v)) -> f t) -> IO t
```

That `forall f. Applicative f => (k -> f (Maybe v)) -> f t` is isomorphic to a [`Bazaar (->)`](https://hackage.haskell.org/package/lens-5.2/docs/Control-Lens-Internal-Bazaar.html#t:Bazaar)

> A Bazaar is like a Traversal that has already been applied to some structure.

This variation makes it very easy to reach into a structure and replace the keys with values in place like this:

``` haskell
tableLookupBatch myTable ([Right ("some metadat", key1), Left "not found", Right ("some other metadata", key2)] & traverse . _Right . _2)
>>> [Right ("some metadat", value1), Left "not found", Right ("some other metadata", value2)]
```

Reply:
I love it. How's this:
```hs
class ReadableTable t k v where
-- ...
    tableLookupBatch' :: t -> (forall f. Applicative f => (k -> f (Maybe v)) -> f r) -> IO r

tableLookupBatch :: (ReadableTable t k v, Each s t' k (Maybe v)) => t -> s -> IO t'
tableLookupBatch t s = tableLookupBatch' t (\k -> each k s)
```

Reply:
Nice touch! So `tableLookupBatch'` gets the more flexible type, but we also expose `tableLookupBatch` which has a more immediately recognizable type.

I wonder how we can improve the naming and avoid the `'` in the names. How about `tableLookupEach` instead of `tableLookupBatch`? It has a similar sense but it has `Each` in the name. 

BTW, `tableLookupEach` exposes a `lens` dependency, but I'm not sure that should be a concern at all. `chainweb-storage` already depends on `lens` and this is a `chainweb`-specific package anyway, so there's very little benefit in being picky here. If we _were_ picky, we could consider a simple `Traversable` constraint, but `Each` is strictly more general, since it doesn't require the container to be polymorphic.

It could also be nice to alias the `forall f. Applicative f => (k -> f (Maybe v)) -> f r` to something that demystifies it. Ed calls it `Bazaar` (behind a `newtype`) as I've mentioned above. The reasoning is that [a "bazaar" holds stores](https://hackage.haskell.org/package/lens-3.9.1/docs/Control-Lens-Internal-Bazaar.html#:~:text=Mnemonically%2C%20a%20Bazaar%20holds%20many%20stores%20and%20you%20can%20easily%20add%20more.), but that sense seems barely relevant here. Maybe we could call it something like `Trail` since it's the type of `traverse someContainer`. So, 

``` haskell
-- | The type of a traversal that was partially applied to a container. This type
-- represents a sequence of locations inside an unspecified container. Transforming
-- the `a`s in each location to `b`s yields a container of type `t`. E.g.
-- flip traverse :: Traversable t => t a -> Trail a b (t b)
-- flip traverse (Just 'a') :: Trail Char b (Maybe b)
-- forOf (_1 . _Just) (Just 'a', 42) :: Trail Char b (Maybe b, Int)
type Trail a b t = forall f. Applicative f => (a -> f b) -> f t
...

tableLookupBatch :: t -> Trail k (Maybe v) t -> IO t
```

Reply:
Something really cool I've noticed about this `Trail`-based interface is that `tableLookupBatch` guarantees that it will provide a `Maybe v` for each `k` in your `Trail`, but the caller of `tableLookupBatch` isn't constrained to retain the shape of the container it's indirectly passing in. For example, you could call it with: 

``` haskell
tableLookupBatch myTable $ flip traverseMaybe myMap 
  where traverseMaybe f = Data.Map.traverseMaybeWithKey (\_ a -> f a)
```

And all the `k`'s that returned a `Nothing` would get removed from `myMap`. So `traverseMaybe` is not really a `Traversal`, but `tableLookupBatch` doesn't care. It's type perfectly captures its contract.